File: web/app/models/invoice.py
"""
发票数据模型
"""
from datetime import datetime
from sqlalchemy import Column, String, Float, DateTime, Enum
import enum

from app.database import Base


class InvoiceStatus(str, enum.Enum):
    PENDING = "pending"
    VERIFIED = "verified"
    REVIEWING = "reviewing"
    FAILED = "failed"


class InvoiceType(str, enum.Enum):
    VAT_SPECIAL = "vat_special"
    VAT_NORMAL = "vat_normal"
    FLIGHT = "flight"
    TAXI = "taxi"
    HOTEL = "hotel"
    OTHER = "other"


class FileType(str, enum.Enum):
    PDF = "pdf"
    JPG = "jpg"
    PNG = "png"
    OFD = "ofd"


class Invoice(Base):
    """发票表"""
    __tablename__ = "invoices"

    id = Column(String(32), primary_key=True, index=True)
    code = Column(String(50), nullable=False, comment="发票代码")
    number = Column(String(50), nullable=False, comment="发票号码")
    type = Column(String(20), default=InvoiceType.OTHER.value, comment="发票类型")
    seller_name = Column(String(200), nullable=False, comment="销方名称")
    buyer_name = Column(String(200), nullable=False, comment="购方名称")
    date = Column(String(20), nullable=False, comment="开票日期")
    amount = Column(Float, default=0.0, comment="金额(不含税)")
    tax_amount = Column(Float, default=0.0, comment="税额")
    total_amount = Column(Float, default=0.0, comment="价税合计")
    status = Column(String(20), default=InvoiceStatus.PENDING.value, comment="状态")
    file_url = Column(String(500), nullable=True, comment="原始文件URL")
    file_type = Column(String(10), default=FileType.PDF.value, comment="文件类型")
    created_at = Column(DateTime, default=datetime.now, comment="创建时间")
    updated_at = Column(DateTime, default=datetime.now, onupdate=datetime.now, comment="更新时间")

File: src/types/invoice.ts
// 发票相关类型定义

/** 发票状态 */
export type InvoiceStatus = 'pending' | 'verified' | 'reviewing' | 'failed'

/** 发票类型 */
export type InvoiceType = 'vat_special' | 'vat_normal' | 'flight' | 'taxi' | 'hotel' | 'other'

/** 发票信息 */
export interface Invoice {
  id: string
  /** 发票代码 */
  code: string
  /** 发票号码 */
  number: string
  /** 发票类型 */
  type: InvoiceType
  /** 销方名称 */
  sellerName: string
  /** 购方名称 */
  buyerName: string
  /** 开票日期 */
  date: string
  /** 金额(不含税) */
  amount: number
  /** 税额 */
  taxAmount: number
  /** 价税合计 */
  totalAmount: number
  /** 状态 */
  status: InvoiceStatus
  /** 原始文件URL */
  fileUrl: string
  /** 文件类型 */
  fileType: 'pdf' | 'jpg' | 'png' | 'ofd'
  /** 创建时间 */
  createdAt: string
  /** 更新时间 */
  updatedAt: string
}

/** 上传文件项 */
export interface UploadFileItem {
  id: string
  fileName: string
  fileSize: number
  fileType: string
  progress: number
  status: 'ready' | 'uploading' | 'success' | 'error'
  file: File
}

/** 合并任务 */
export interface MergeTask {
  id: string
  invoiceIds: string[]
  status: 'pending' | 'processing' | 'completed' | 'failed'
  outputType: 'pdf' | 'zip'
  totalPages: number
  totalAmount: number
  createdAt: string
  downloadUrl?: string
}

/** 统计数据 */
export interface DashboardStats {
  /** 本月已处理数量 */
  processedCount: number
  /** 处理数量变化百分比 */
  processedChange: number
  /** 待审核任务数 */
  pendingCount: number
  /** 待审核变化百分比 */
  pendingChange: number
  /** 本月节省税额 */
  savedTax: number
  /** 节省变化百分比 */
  savedChange: number
}

/** 分页请求 */
export interface PageRequest {
  page: number
  pageSize: number
}

/** 分页响应 */
export interface PageResponse<T> {
  data: T[]
  total: number
  page: number
  pageSize: number
}

/** API响应 */
export interface ApiResponse<T> {
  code: number
  message: string
  data: T
}
